### ISSUE TITLE
PDF is still quite large after running Document.select() and selecting only a few pages

### ISSUE THREAD

**Issue by MerlijnWajer** [pymupdf/PyMuPDF#1061]
## Describe the bug (mandatory)

To be perfectly clear, I am not entirely sure if this is a bug, but was hoping to get your take.
We have code that takes a full (book) PDF and selects only a specific chapter and produces a PDF of that. It's built around `Document.select` of PymuPDF, and saved with `Document.save(garbage=2)`. In general, it works quite smooth, and produces small PDFs (only the pages that are supposed to kept are actually kept), but with a particular PDF this doesn't seem to happen.
Perhaps the PDF has a weird setup where each page references each other or something, but that wasn't immediately obvious to me.

The input PDF is 86MB and calling `Document.select([1])` (or any other page) still creates a 105MB PDF.

## To Reproduce (mandatory)

PDF in question: https://archive.org/~merlijn/encyclopediafood00caba.pdf

Code:

```
$ cat minsplit.py
import sys
import fitz

doc = fitz.open(sys.argv[1])
doc.select([1])
doc.save(sys.argv[2], garbage=2)
```

Use like:

```
$ python minsplit.py /tmp/encyclopediafood00caba.pdf /tmp/out.pdf
$ ls -lsh /tmp/encyclopediafood00caba.pdf /tmp/out.pdf
 86M -rw-r--r-- 1 merlijn merlijn  86M May 26 21:19 /tmp/encyclopediafood00caba.pdf
105M -rw-r--r-- 1 merlijn merlijn 105M May 26 21:58 /tmp/out.pdf
```

Compare to a PDF without problems:

```
$ python minsplit.py /tmp/lanjingdeyanjing0000bing.pdf /tmp/out.pdf
$ ls -lsh /tmp/lanjingdeyanjing0000bing.pdf /tmp/out.pdf
 15M -rw-r--r-- 1 merlijn merlijn  15M May 26 21:35 /tmp/lanjingdeyanjing0000bing.pdf
728K -rw-r--r-- 1 merlijn merlijn 725K May 26 21:59 /tmp/out.pdf
```

## Expected behavior (optional)

The resulting document would be smaller than 105MB.

## Your configuration (mandatory)

```
3.8.9 (default, May 16 2021, 12:11:24)
[GCC 9.2.0]
 linux

PyMuPDF 1.18.14: Python bindings for the MuPDF 1.18.0 library.
Version date: 2021-05-17 19:46:05.
Built for Python 3.8 on linux (64-bit).
```

## Additional context (optional)

When using `Document.save(garbage=4)`, it's also really slow. I imagine this is related to the problem as reported because it deduplicates all the streams that should not be saved to begin with.

In general `Document.split()` works great. :-)

**Comment by JorjMcKie** [comment]
I see.
I am still in the process of downloading the example PDF (sloooow Internet). In the meantime a few comments.
Your intention can be achieved in more than one way:
1. ``select()``
2. ``delete_pages()`` using a sequence of page numbers representing the set theoretic complement of the sequence used in `select()`. This is a new feature support in v1.18.13 (or .14?).
3. Make a new PDF and insert the original's pages.

Whiel 1. and 2. should lead to the same result, alternative 3. will in general be smaller, because only the pages and their (recursive) dependents are copied over - i.e. no page labels, no embedded files.

I have seen "resistant" PDF examples that simply wouldn't get smaller in any of the 3 ways. After some investigation, it turned out, that the reason was a `/B` key in the page definition, which (across some intermediate levels) finally referenced **_all_** other pages of the PDF. So the result had a page count of 1, but still contained **_everything_** of the old file.
Atfer remove the `/B` from the page object(s), everything was fine.
So I am halfway certain, we have a similar situation here ... but let's see.

**Comment by MerlijnWajer** [comment]
Thanks for the quick response ( as always :-)  )

A few remarks:

* Preserving the page labels and page text is actually a plus for this use case
* I wrote my own code earlier using `delete_page`, but ran into inexplicable problems in Adobe PDF (and only in Adobe PDF, all other viewers were just fine) that complained about a corrupt PDF and I was not able to figure it out. Then realised there was a cool `Document.select()` method and didn't look back since. I don't recall if I filed a bug for that at the time, but if you'd like I can try to reproduce that problem as well.

**Comment by JorjMcKie** [comment]
Aha - it **_does_** contain a ``/B`` key:
![grafik](https://user-images.githubusercontent.com/8290722/119729770-0a4f7d80-be43-11eb-94d7-e07dad2a9205.png)

The PDF also contains `/StructParents` keys in the page definitions - which lead to a plethora of cris-cross references throughout the file - similar to the `/B`.
![grafik](https://user-images.githubusercontent.com/8290722/119734358-e0995500-be48-11eb-8215-b97e0e6c981d.png)

So using `delete_pages` will indeed not help here and lead to the same result as select.
When looking in the resulting (select of delete_pages used) PDF, I see several old page objects are still there (not adressable as pages anymore though). Obviously a result of the StructParent structure of the PDF.

As you indicated: this PDF simply is a bad candidate for shrinking its size down by deleting pages ... there are too many internal references back and forth which prevent objects to become unreferenced.

What really helps here, is using `insert_pdf()`. The result is just 41 KB large - but of course there are losses which you might perceive as prohibitive.

Bottom line - as you already guessed - it is not a bug.

**Comment by MerlijnWajer** [comment]
I suppose if I don't care too much about the Accessibility stuff, I could get rid of the StructParents and B key if they exist. We can probably close this. Thank you for the clear explanation!

**Comment by JorjMcKie** [comment]
> get rid of the StructParents and B key if they exist

You could do this to achieve it:

```python
doc.xref_set_key(page.xref, "B", "null")
doc.xref_set_key(page.xref, "StructParents", "null")
```
